contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,794 @@
+
+When starting the device plugin is expected to make a (client) gRPC call
+to the `Register` function that Kubelet exposes.
+
+The communication between Kubelet is expected to happen only through Unix
+sockets and follow this simple pattern:
+1. The device plugins starts it's gRPC server
+2. The device plugins sends a `RegisterRequest` to Kubelet (through a
+   gRPC request)
+4. Kubelet starts it's Discovery phase and calls `ListDevices` and `HealthCheck`
+5. Kubelet answers to the `RegisterRequest` with a `RegisterResponse`
+   containing any error Kubelet might have encountered
+
+### Unix Socket
+
+Device Plugins are expected to communicate with Kubelet through gRPC
+on an Unix socket.
+When starting the gRPC server, they are expected to create a unix socket
+at the following host path: `/var/run/kubernetes`.
+
+For non bare metal device plugin this means they will have to mount the folder
+as a volume in their pod spec ([see Installation](##installation)).
+
+Device plugins can expect to find the socket to register themselves on
+the host at the following path:
+`/var/run/Kubernetes/kubelet.sock`.
+
+### Protocol Overview
+
+When first registering themselves against Kubelet, the device plugin
+will send:
+  * The name of their unix socket
+  * [The API version against which they were built](#versioning).
+  * Their `Vendor` ID or name of the device plugin
+
+Kubelet answers with the minimum version it supports and whether or
+not there was an error. The errors may include (but not limited to):
+  * API version not supported
+  * A device plugin was already registered for this vendor
+  * A device plugin already registered this device
+  * Vendor is not consistent across discovered devices
+
+Kubelet will then interact with the plugin through the following functions:
+  * `ListDevices`: List Devices
+  * `HealthCheck`: Returns a stream that is written to when a Device becomes
+     unhealty
+  * `Allocate`: Called when creating a container with a list of devices
+
+The device plugin is also expected to periodically call the `Heartbeat` function
+exposed by Kubelet and issue a `Registration` request when it either can't reach
+Kubelet or Kubelet answers with a `KO` response.
+
+![Process](device-plugin.png)
+
+
+### Protobuf specification
+
+```go
+service PluginRegistration {
+	rpc Register(RegisterRequest) returns (RegisterResponse) {}
+	rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse) {}
+}
+
+service DevicePlugin {
+	rpc ListDevices(Empty) returns (stream Device) {}
+	rpc HealthCheck(Empty) returns (stream Device) {}
+
+	rpc Allocate(AllocateRequest) returns (AllocateResponse) {}
+}
+
+message RegisterRequest {
+	// Version of the API the Device Plugin was built against
+	string version = 1;
+	// Name of the unix socket the device plugin is listening on
+	string unixsocket = 2;
+	// Name of the devices the device plugin wants to register
+	// A device plugin can only register one kind of devices
+	string vendor = 3;
```

Comment:
If it's the name of the device, why not name the field `DeviceName` or something like that?